cmake/libxmp-checks.cmake

```diff
@@ -0,0 +1,194 @@
+    cmake_policy(SET CMP0075 NEW)
+endif()
+
+if(POLICY CMP0077) # Cache variables override since 3.12
+    cmake_policy(SET CMP0077 NEW)
+endif()
+
+
+# ======== Compiler ========
+
+if(CMAKE_C_COMPILER_ID MATCHES "Clang")
+    set(HAVE_CLANG TRUE)
+elseif(CMAKE_COMPILER_IS_GNUCC)
+    set(HAVE_GCC TRUE)
+endif()
+
+message("-- Check for working linker: ${CMAKE_LINKER}")
+execute_process(COMMAND ${CMAKE_LINKER} -v
+                OUTPUT_VARIABLE LINKER_OUTPUT
+                ERROR_VARIABLE  LINKER_OUTPUT)
+
+string(REGEX REPLACE "[\r\n].*" "" LINKER_OUTPUT_LINE "${LINKER_OUTPUT}")
+message(STATUS "Linker identification: ${LINKER_OUTPUT_LINE}")
+
+if("${LINKER_OUTPUT}" MATCHES ".*GNU.*" OR "${LINKER_OUTPUT}" MATCHES ".*with BFD.*")
+    set(HAVE_GNU_LD TRUE)
+endif()
+
+
+# ======== Includes ========
+
+include(CheckSymbolExists)
+include(CheckFunctionExists)
+include(CheckIncludeFile)
+include(CMakePushCheckState)
+include(CheckCSourceCompiles)
+include(CheckCCompilerFlag)
+include(TestBigEndian)
+
+# If platform is Emscripten
+if(${CMAKE_SYSTEM_NAME} STREQUAL "Emscripten")
+    set(EMSCRIPTEN 1)
+endif()
+
+
+# ======== Macros ========
+macro(xmp_add_warning_flag WARNINGFLAG WARNING_VAR)
+    check_c_compiler_flag("${WARNINGFLAG}" HAVE_W_${WARNING_VAR})
+    if(HAVE_W_${WARNING_VAR})
+       set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${WARNINGFLAG}")
+    endif()
+endmacro()
+
+macro(xmp_disable_warning_flag WARNINGFLAG WARNING_VAR)
+    check_c_compiler_flag("-W${WARNINGFLAG}" HAVE_W_${WARNING_VAR})
+    if(HAVE_W_${WARNING_VAR})
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-${WARNINGFLAG}")
+    endif()
+endmacro()
+
+macro(xmp_check_function SYMBOL HEADER VARNAME)
+    check_function_exists(${SYMBOL} ${VARNAME}_F)
+    check_symbol_exists(${SYMBOL} "${HEADER}" ${VARNAME})
+    if(${VARNAME} AND ${VARNAME}_F)
+        add_definitions(-D${VARNAME}=1)
+    endif()
+endmacro()
+
+macro(xmp_check_include HEADER VARNAME)
+    check_include_file(${HEADER} ${VARNAME})
+    if(${VARNAME})
+        add_definitions(-D${VARNAME}=1)
+    endif()
+endmacro()
+
+
+# ======== Flags and warning ========
+
+string(TOLOWER "${CMAKE_BUILD_TYPE}" CMAKE_BUILD_TYPE_LOWER)
+
+set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -DNDEBUG")
+set(CMAKE_C_FLAGS_MINSIZEREL "${CMAKE_C_FLAGS_MINSIZEREL} -DNDEBUG")
+set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -DDEBUG -D_DEBUG")
+
+if(MSVC)
+    # Force to always compile with W4
+    if(CMAKE_C_FLAGS MATCHES "/W[0-4]")
+        string(REGEX REPLACE "/W[0-4]" "/W4" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
+    else()
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W4")
+    endif()
+    # Disable bogus MSVC warnings
+    add_definitions(-D_CRT_SECURE_NO_WARNINGS)
+    # Tune warnings
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /wd4244 /wd4018 /wd4996 /wd4048 /wd4267 /wd4127")
+else()
+    xmp_add_warning_flag("-Wall" GCC_WALL)
+    xmp_add_warning_flag("-Wwrite-strings" WRITE_STRINGS)
```

Comment:
This literally just got merged (sorry for making more work for you...), but the Makefile system now adds `-Wextra -Wno-unused-parameters -Wno-sign-compare` when all three are available.

Reply:
Done

Reply:
> edit: this was also added to lite and test-dev, if that's relevant (not sure how much this gets reused).

Adding these flags at the parent project, all sub-directory projects will inherit this.